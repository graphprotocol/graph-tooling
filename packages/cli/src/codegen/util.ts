export function disambiguateNames<T>({
  values,
  getName,
  setName,
}: {
  values: T[];
  getName: (value: T, index: number) => string;
  setName: (value: T, name: string) => void;
}) {
  const collisionCounter = new Map();
  return values.map((value, index) => {
    const name = getName(value, index);
    const counter = collisionCounter.get(name);
    if (counter === undefined) {
      collisionCounter.set(name, 1);
      return setName(value, name);
    }
    collisionCounter.set(name, counter + 1);
    return setName(value, `${name}${counter}`);
  });
}

export function isTupleType(t: string) {
  return t === 'tuple';
}

export function containsTupleType(t: string) {
  return isTupleType(t) || isTupleArrayType(t) || isTupleMatrixType(t);
}

export function isTupleArrayType(t: string) {
  return t.match(/^tuple\[([0-9]+)?\]$/);
}

export function isTupleMatrixType(t: string) {
  return t.match(/^tuple\[([0-9]+)?\]\[([0-9]+)?\]$/);
}

export const unrollTuple = ({
  path,
  value,
}: {
  path: string[];
  index: number; // TODO: index is unused, do we really need it?
  value: any;
}) =>
  value.components.reduce((acc: any[], component: any, index: number) => {
    const name = component.name || `value${index}`;
    return acc.concat(
      component.type === 'tuple'
        ? unrollTuple({ path: [...path, name], index, value: component })
        : [{ path: [...path, name], type: component.type }],
    );
  }, []);
