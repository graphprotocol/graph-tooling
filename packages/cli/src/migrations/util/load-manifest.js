const fs = require('fs-extra')
const path = require('path')
const yaml = require('js-yaml')

async function loadManifest(manifestFile) {
  if(manifestFile.match(/.js$/)) {
    return require(path.resolve(manifestFile))
  }
  else {
    return yaml.safeLoad(await fs.readFile(manifestFile, 'utf-8'))
  }
}

module.exports = {
  loadManifest,
}
