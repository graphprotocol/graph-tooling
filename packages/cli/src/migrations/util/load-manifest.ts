import path from 'path';
import fs from 'fs-extra';
import yaml from 'js-yaml';

export async function loadManifest(manifestFile: string) {
  if (manifestFile.match(/.js$/)) {
    return require(path.resolve(manifestFile));
  }
  return yaml.safeLoad(await fs.readFile(manifestFile, 'utf-8'));
}
