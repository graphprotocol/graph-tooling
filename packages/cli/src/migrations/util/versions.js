const path = require('path')
const fs = require('fs-extra')
const yaml = require('js-yaml')

const getGraphTsVersion = async sourceDir => {
  let graphTsPath

  for (
    let dir = path.resolve(sourceDir);
    // Terminate after the root dir or when we have found node_modules
    dir !== undefined;
    // Continue with the parent directory, terminate after the root dir
    dir = path.dirname(dir) === dir ? undefined : path.dirname(dir)
  ) {
    let graphTsNodeModulesPath = path.join(
      dir,
      'node_modules',
      '@graphprotocol',
      'graph-ts',
    )

    if (fs.existsSync(graphTsNodeModulesPath)) {
      graphTsPath = graphTsNodeModulesPath
      // Loop until we find the first occurrence of graph-ts in node_modules
      break
    }
  }

  let pkgJsonFile = path.join(
    graphTsPath,
    'package.json',
  )
  let data = await fs.readFile(pkgJsonFile)
  let jsonData = JSON.parse(data)
  return jsonData.version
}

module.exports = {
  getGraphTsVersion,
}
