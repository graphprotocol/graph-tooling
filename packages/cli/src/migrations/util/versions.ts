import path from 'path';
import fs from 'fs-extra';

export async function getGraphTsVersion(sourceDir: string) {
  let graphTsPath!: string;

  for (
    let dir: string | undefined = path.resolve(sourceDir);
    // Terminate after the root dir or when we have found node_modules
    dir !== undefined;
    // Continue with the parent directory, terminate after the root dir
    dir = path.dirname(dir) === dir ? undefined : path.dirname(dir)
  ) {
    const graphTsNodeModulesPath = path.join(dir, 'node_modules', '@graphprotocol', 'graph-ts');

    if (fs.existsSync(graphTsNodeModulesPath)) {
      graphTsPath = graphTsNodeModulesPath;
      // Loop until we find the first occurrence of graph-ts in node_modules
      break;
    }
  }

  const pkgJsonFile = path.join(graphTsPath, 'package.json');
  const data = await fs.readFile(pkgJsonFile);
  const jsonData = JSON.parse(data.toString());
  return jsonData.version;
}
