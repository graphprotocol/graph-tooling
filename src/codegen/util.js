const disambiguateNames = ({ values, getName, setName }) => {
  let collisionCounter = new Map()
  return values.map((value, index) => {
    let name = getName(value, index)
    let counter = collisionCounter.get(name)
    if (counter === undefined) {
      collisionCounter.set(name, 1)
      return setName(value, name)
    } else {
      collisionCounter.set(name, counter + 1)
      return setName(value, `${name}${counter}`)
    }
  })
}

const isTupleType = t => {
  return t === 'tuple'
}

const containsTupleType = t => {
  return isTupleType(t) || isTupleArrayType(t)
}

const isTupleArrayType = t => {
  return t.match(/^tuple\[([0-9]+)?\]$/)
}

const unrollTuple = ({ path, index, value }) =>
  value.components.reduce((acc, component, index) => {
    let name = component.name || `value${index}`
    return acc.concat(
      component.type === 'tuple'
        ? unrollTuple({
            path: [...path, name],
            index,
            value: component,
          })
        : [{ path: [...path, name], type: component.type }],
    )
  }, [])

module.exports = {
  containsTupleType,
  disambiguateNames,
  isTupleType,
  isTupleArrayType,
  unrollTuple,
}
