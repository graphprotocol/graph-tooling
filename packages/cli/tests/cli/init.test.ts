import path from 'path';
import { cliTest } from './util';

describe('Init', () => {
  const baseDir = path.join(__dirname, 'init');

  describe('Ethereum', () => {
    const ethereumBaseDir = path.join(baseDir, 'ethereum');

    cliTest(
      'From example',
      [
        'init',
        '--protocol',
        'ethereum',
        '--studio',
        '--from-example',
        'ethereum/gravatar',
        'user/example-subgraph',
        path.join(ethereumBaseDir, 'from-example'),
      ],
      path.join('init', 'ethereum', 'from-example'),
      {
        exitCode: 0,
        timeout: 100_000,
        cwd: ethereumBaseDir,
        deleteDir: true,
      },
    );

    cliTest(
      'From contract',
      [
        'init',
        '--protocol',
        'ethereum',
        '--studio',
        '--from-contract',
        '0xF87E31492Faf9A91B02Ee0dEAAd50d51d56D5d4d',
        '--network',
        'mainnet',
        'user/subgraph-from-contract',
        path.join(ethereumBaseDir, 'from-contract'),
      ],
      path.join('init', 'ethereum', 'from-contract'),
      {
        exitCode: 0,
        timeout: 100_000,
        cwd: ethereumBaseDir,
        deleteDir: true,
      },
    );

    cliTest(
      'From contract with abi',
      [
        'init',
        '--protocol',
        'ethereum',
        '--studio',
        '--from-contract',
        '0xF87E31492Faf9A91B02Ee0dEAAd50d51d56D5d4d',
        '--abi',
        path.join(ethereumBaseDir, 'abis', 'Marketplace.json'),
        '--network',
        'mainnet',
        'user/subgraph-from-contract-with-abi',
        path.join(ethereumBaseDir, 'from-contract-with-abi'),
      ],
      path.join('init', 'ethereum', 'from-contract-with-abi'),
      {
        exitCode: 0,
        timeout: 100_000,
        cwd: ethereumBaseDir,
        deleteDir: true,
      },
    );

    cliTest(
      'From contract with abi and structs',
      [
        'init',
        '--protocol',
        'ethereum',
        '--studio',
        '--from-contract',
        '0x1E0447b19BB6EcFdAe1e4AE1694b0C3659614e4e',
        '--abi',
        path.join(ethereumBaseDir, 'abis', 'SoloMargin.json'),
        '--network',
        'mainnet',
        'user/subgraph-from-contract-with-abi-and-structs',
        path.join(ethereumBaseDir, 'from-contract-with-abi-and-structs'),
      ],
      path.join('init', 'ethereum', 'from-contract-with-abi-and-structs'),
      {
        exitCode: 0,
        timeout: 100_000,
        cwd: ethereumBaseDir,
        deleteDir: true,
      },
    );

    cliTest(
      'From contract with list items in abi',
      [
        'init',
        '--protocol',
        'ethereum',
        '--studio',
        '--from-contract',
        '0x1E0447b19BB6EcFdAe1e4AE1694b0C3659614e4e',
        '--abi',
        path.join(ethereumBaseDir, 'abis', 'Airdropped.json'),
        '--network',
        'mainnet',
        'user/subgraph-from-contract-with-lists-in-abi',
        path.join(ethereumBaseDir, 'from-contract-with-lists-in-abi'),
      ],
      path.join('init', 'ethereum', 'from-contract-with-lists-in-abi'),
      {
        exitCode: 0,
        timeout: 100_000,
        cwd: ethereumBaseDir,
        deleteDir: true,
      },
    );

    cliTest(
      'From contract with overloaded elements',
      [
        'init',
        '--protocol',
        'ethereum',
        '--studio',
        '--from-contract',
        '0x1E0447b19BB6EcFdAe1e4AE1694b0C3659614e4e',
        '--abi',
        path.join(ethereumBaseDir, 'abis', 'OverloadedElements.json'),
        '--network',
        'mainnet',
        'user/subgraph-from-contract-with-overloaded-elements',
        path.join(ethereumBaseDir, 'from-contract-with-overloaded-elements'),
      ],
      path.join('init', 'ethereum', 'from-contract-with-overloaded-elements'),
      {
        exitCode: 0,
        timeout: 100_000,
        cwd: ethereumBaseDir,
        deleteDir: true,
      },
    );

    cliTest(
      'From contract with index events and abi with ID in events',
      [
        'init',
        '--protocol',
        'ethereum',
        '--studio',
        '--from-contract',
        '0x57f1887a8bf19b14fc0df6fd9b2acc9af147ea85',
        '--index-events',
        '--abi',
        path.join(ethereumBaseDir, 'abis', 'Ids.json'),
        '--network',
        'mainnet',
        'user/subgraph-from-contract-with-index-events-and-abi-with-id',
        path.join(ethereumBaseDir, 'duplicate-ids'),
      ],
      path.join('init', 'ethereum', 'duplicate-ids'),
      {
        exitCode: 0,
        timeout: 100_000,
        cwd: ethereumBaseDir,
        deleteDir: true,
        runBuild: true,
      },
    );
  });

  describe('NEAR', () => {
    const nearBaseDir = path.join(baseDir, 'near');

    cliTest(
      'From contract',
      [
        'init',
        '--protocol',
        'near',
        '--product',
        'hosted-service',
        '--from-contract',
        'app.good-morning.near',
        '--network',
        'near-mainnet',
        'user/near-from-contract',
        path.join(nearBaseDir, 'from-contract'),
      ],
      path.join('init', 'near', 'from-contract'),
      {
        exitCode: 0,
        timeout: 100_000,
        cwd: nearBaseDir,
        deleteDir: true,
      },
    );
  });
});
