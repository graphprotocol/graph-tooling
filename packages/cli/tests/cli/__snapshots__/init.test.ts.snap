// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Init > Ethereum > From contract 1`] = `
" ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will stop initializing a Git repository.
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
- Fetching ABI from Etherscan
✔ Fetching ABI from Etherscan
- Create subgraph scaffold
  Generate subgraph
- Create subgraph scaffold
  Write subgraph to directory
- Create subgraph scaffold
✔ Create subgraph scaffold
- Initialize networks config
✔ Initialize networks config
- Install dependencies with yarn
✔ Install dependencies with yarn
- Generate ABI and schema types with yarn codegen
✔ Generate ABI and schema types with yarn codegen
"
`;

exports[`Init > Ethereum > From contract 2`] = `0`;

exports[`Init > Ethereum > From contract 3`] = `
"
Subgraph user/subgraph-from-contract created in from-contract

Next steps:

  1. Run \`graph auth\` to authenticate with your deploy key.

  2. Type \`cd from-contract\` to enter the subgraph.

  3. Run \`yarn deploy\` to deploy the subgraph.

Make sure to visit the documentation on https://thegraph.com/docs/ for further information.
"
`;

exports[`Init > Ethereum > From contract with abi 1`] = `
" ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will stop initializing a Git repository.
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
- Create subgraph scaffold
  Generate subgraph
- Create subgraph scaffold
  Write subgraph to directory
- Create subgraph scaffold
✔ Create subgraph scaffold
- Initialize networks config
✔ Initialize networks config
- Install dependencies with yarn
✔ Install dependencies with yarn
- Generate ABI and schema types with yarn codegen
✔ Generate ABI and schema types with yarn codegen
"
`;

exports[`Init > Ethereum > From contract with abi 2`] = `0`;

exports[`Init > Ethereum > From contract with abi 3`] = `
"
Subgraph user/subgraph-from-contract-with-abi created in from-contract-with-abi

Next steps:

  1. Run \`graph auth\` to authenticate with your deploy key.

  2. Type \`cd from-contract-with-abi\` to enter the subgraph.

  3. Run \`yarn deploy\` to deploy the subgraph.

Make sure to visit the documentation on https://thegraph.com/docs/ for further information.
"
`;

exports[`Init > Ethereum > From contract with abi and structs 1`] = `
" ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will stop initializing a Git repository.
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
- Create subgraph scaffold
  Generate subgraph
- Create subgraph scaffold
  Write subgraph to directory
- Create subgraph scaffold
✔ Create subgraph scaffold
- Initialize networks config
✔ Initialize networks config
- Install dependencies with yarn
✔ Install dependencies with yarn
- Generate ABI and schema types with yarn codegen
✔ Generate ABI and schema types with yarn codegen
"
`;

exports[`Init > Ethereum > From contract with abi and structs 2`] = `0`;

exports[`Init > Ethereum > From contract with abi and structs 3`] = `
"
Subgraph user/subgraph-from-contract-with-abi-and-structs created in from-contract-with-abi-and-structs

Next steps:

  1. Run \`graph auth\` to authenticate with your deploy key.

  2. Type \`cd from-contract-with-abi-and-structs\` to enter the subgraph.

  3. Run \`yarn deploy\` to deploy the subgraph.

Make sure to visit the documentation on https://thegraph.com/docs/ for further information.
"
`;

exports[`Init > Ethereum > From contract with index events and abi with ID in events 1`] = `
" ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will stop initializing a Git repository.
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
- Create subgraph scaffold
  Generate subgraph
- Create subgraph scaffold
  Write subgraph to directory
- Create subgraph scaffold
✔ Create subgraph scaffold
- Initialize networks config
✔ Initialize networks config
- Install dependencies with yarn
✔ Install dependencies with yarn
- Generate ABI and schema types with yarn codegen
✔ Generate ABI and schema types with yarn codegen
"
`;

exports[`Init > Ethereum > From contract with index events and abi with ID in events 2`] = `0`;

exports[`Init > Ethereum > From contract with index events and abi with ID in events 3`] = `
"
Subgraph user/subgraph-from-contract-with-index-events-and-abi-with-id created in duplicate-ids

Next steps:

  1. Run \`graph auth\` to authenticate with your deploy key.

  2. Type \`cd duplicate-ids\` to enter the subgraph.

  3. Run \`yarn deploy\` to deploy the subgraph.

Make sure to visit the documentation on https://thegraph.com/docs/ for further information.
"
`;

exports[`Init > Ethereum > From contract with list items in abi 1`] = `
" ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will stop initializing a Git repository.
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
- Create subgraph scaffold
  Generate subgraph
- Create subgraph scaffold
  Write subgraph to directory
- Create subgraph scaffold
✔ Create subgraph scaffold
- Initialize networks config
✔ Initialize networks config
- Install dependencies with yarn
✔ Install dependencies with yarn
- Generate ABI and schema types with yarn codegen
✔ Generate ABI and schema types with yarn codegen
"
`;

exports[`Init > Ethereum > From contract with list items in abi 2`] = `0`;

exports[`Init > Ethereum > From contract with list items in abi 3`] = `
"
Subgraph user/subgraph-from-contract-with-lists-in-abi created in from-contract-with-lists-in-abi

Next steps:

  1. Run \`graph auth\` to authenticate with your deploy key.

  2. Type \`cd from-contract-with-lists-in-abi\` to enter the subgraph.

  3. Run \`yarn deploy\` to deploy the subgraph.

Make sure to visit the documentation on https://thegraph.com/docs/ for further information.
"
`;

exports[`Init > Ethereum > From contract with overloaded elements 1`] = `
" ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will stop initializing a Git repository.
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
- Create subgraph scaffold
  Generate subgraph
- Create subgraph scaffold
  Write subgraph to directory
- Create subgraph scaffold
✔ Create subgraph scaffold
- Initialize networks config
✔ Initialize networks config
- Install dependencies with yarn
✔ Install dependencies with yarn
- Generate ABI and schema types with yarn codegen
✔ Generate ABI and schema types with yarn codegen
"
`;

exports[`Init > Ethereum > From contract with overloaded elements 2`] = `0`;

exports[`Init > Ethereum > From contract with overloaded elements 3`] = `
"
Subgraph user/subgraph-from-contract-with-overloaded-elements created in from-contract-with-overloaded-elements

Next steps:

  1. Run \`graph auth\` to authenticate with your deploy key.

  2. Type \`cd from-contract-with-overloaded-elements\` to enter the subgraph.

  3. Run \`yarn deploy\` to deploy the subgraph.

Make sure to visit the documentation on https://thegraph.com/docs/ for further information.
"
`;

exports[`Init > Ethereum > From example 1`] = `
" ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will stop initializing a Git repository.
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
- Cloning example subgraph
✔ Cloning example subgraph
- Initialize networks config
✔ Initialize networks config
- Update subgraph name and commands in package.json
✔ Update subgraph name and commands in package.json
- Install dependencies with yarn
✔ Install dependencies with yarn
- Generate ABI and schema types with yarn codegen
✔ Generate ABI and schema types with yarn codegen
"
`;

exports[`Init > Ethereum > From example 2`] = `0`;

exports[`Init > Ethereum > From example 3`] = `
"
Subgraph user/example-subgraph created in from-example

Next steps:

  1. Run \`graph auth\` to authenticate with your deploy key.

  2. Type \`cd from-example\` to enter the subgraph.

  3. Run \`yarn deploy\` to deploy the subgraph.

Make sure to visit the documentation on https://thegraph.com/docs/ for further information.
"
`;

exports[`Init > NEAR > From contract 1`] = `
" ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will stop initializing a Git repository.
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
 ›   Warning: In next major version, this flag will be removed. By default we 
 ›   will deploy to the Graph Studio. Learn more about Sunrise of Decentralized
 ›    Data 
 ›   https://thegraph.com/blog/unveiling-updated-sunrise-decentralized-data/
- Create subgraph scaffold
  Generate subgraph
- Create subgraph scaffold
  Write subgraph to directory
- Create subgraph scaffold
✔ Create subgraph scaffold
- Initialize networks config
✔ Initialize networks config
- Install dependencies with yarn
✔ Install dependencies with yarn
- Generate ABI and schema types with yarn codegen
✔ Generate ABI and schema types with yarn codegen
"
`;

exports[`Init > NEAR > From contract 2`] = `0`;

exports[`Init > NEAR > From contract 3`] = `
"
Subgraph user/near-from-contract created in from-contract

Next steps:

  1. Run \`graph auth\` to authenticate with your deploy key.

  2. Type \`cd from-contract\` to enter the subgraph.

  3. Run \`yarn deploy\` to deploy the subgraph.

Make sure to visit the documentation on https://thegraph.com/docs/ for further information.
"
`;
